bowjson.go

```diff
@@ -1,88 +1,173 @@
 package bow
 
 import (
+	"bytes"
 	"encoding/json"
 	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
 )
 
 func (b *bow) SetMarshalJSONRowBased(rowOriented bool) {
 	b.marshalJSONRowBased = rowOriented
 }
 
-type jsonColSchema struct {
+var (
+	ErrUnmarshalJSON = errors.New("could not unmarshal JSON to bow")
+	ErrEncodeBow     = errors.New("could not encode bow to JSON body")
+	ErrDecodeJSON    = errors.New("could not decode JSON to bow")
+)
+
+type jsonField struct {
 	Name string `json:"name"`
 	Type string `json:"type"`
 }
 
+type jsonSchema struct {
+	Fields []jsonField `json:"fields"`
+}
+
 type jsonRecord struct {
-	Schema struct {
-		Fields []jsonColSchema `json:"fields"`
-	} `json:"schema"`
-	Data []map[string]interface{} `json:"data"`
+	Schema jsonSchema               `json:"schema"`
+	Data   []map[string]interface{} `json:"data"`
 }
 
 func (b *bow) MarshalJSON() ([]byte, error) {
 	if !b.marshalJSONRowBased {
-		// it will be handled natively by arrow, today (24 oct 2019) still in arrow's internal packages
+		// it will be handled natively by arrow, today (26 nov 2020) still in arrow's internal packages
 		panic("bow: column based json marshaller not implemented")
 	}
+
 	rowBased := jsonRecord{}
 	for _, col := range b.Schema().Fields() {
-		rowBased.Schema.Fields = append(rowBased.Schema.Fields, jsonColSchema{
+		rowBased.Schema.Fields = append(rowBased.Schema.Fields, jsonField{
 			Name: col.Name,
 			Type: col.Type.Name(),
 		})
 	}
+
 	for row := range b.RowMapIter() {
 		if len(row) == 0 {
 			continue
 		}
 		rowBased.Data = append(rowBased.Data, row)
 	}
+
 	return json.Marshal(rowBased)
 }
 
-func (b *bow) UnmarshalJSON(data []byte) error {
-	jsonBow := jsonRecord{}
-	if err := json.Unmarshal(data, &jsonBow); err != nil {
-		return err
+func UnmarshalJSON(data []byte) (Bow, error) {
```

Comment:
this must be a bow method otherwise embedding in other struct will not work. I would expect users to be able to do:
```go
var payload = struct {
  Bow  bow.Bow
  OtherStuff string
}{
  Bow:   bow.New(),
  OtherStuff: "",
}
err := json.Unmarshall(data, &payload)
panic(err)
// payload contain a valid bow.
```
an example should present the use case.